File: leetcode/201-300/296-best-meeting-point/bench/meetpoint.py
"""Benchmark mirror of meetpoint.kara - LeetCode #296, separable medians.

Same two scans (row-major then column-major), same reused scratch, same sink.
Written as explicit index loops rather than comprehensions so it times the same
algorithm the other four mirrors run, not Python's fastest way to say it.
"""

NCASES = 400
DIM = 128
PASSES = 30
CELLS = DIM * DIM
MOD = 1000000007


def main() -> None:
    corpus = bytearray(NCASES * CELLS)
    state = 24601
    for n in range(NCASES * CELLS):
        state = (state * 1103515245 + 12345) & 0x7FFFFFFF
        if (state // 65536) % 100 < 10:
            corpus[n] = 1

    rows = [0] * CELLS
    cols = [0] * CELLS
    checksum = 0

    for _ in range(PASSES):
        for ci in range(NCASES):
            base = ci * CELLS

            k = 0
            for r in range(DIM):
                row_base = base + r * DIM
                for c in range(DIM):
                    if corpus[row_base + c] == 1:
                        rows[k] = r
                        k += 1

            k2 = 0
            for c in range(DIM):
                for r in range(DIM):
                    if corpus[base + r * DIM + c] == 1:
                        cols[k2] = c
                        k2 += 1

            total = 0
            if k > 0:
                mr = rows[k // 2]
                mc = cols[k // 2]
                for i in range(k):
                    dr = rows[i] - mr
                    total += -dr if dr < 0 else dr
                    dc = cols[i] - mc
                    total += -dc if dc < 0 else dc

            checksum = (checksum + total) % MOD

    print(f"checksum {checksum}")


if __name__ == "__main__":
    main()

File: leetcode/201-300/296-best-meeting-point/bench/meetpoint.rs
// Benchmark mirror of meetpoint.kara — LeetCode #296, separable medians.
// Same two scans (row-major then column-major), same reused scratch, same sink.

const NCASES: usize = 400;
const DIM: usize = 128;
const PASSES: usize = 30;
const CELLS: usize = DIM * DIM;
const MOD: i64 = 1_000_000_007;

fn main() {
    let mut corpus = vec![0u8; NCASES * CELLS];
    let mut state: i64 = 24601;
    for slot in corpus.iter_mut() {
        state = (state * 1103515245 + 12345) & 0x7fff_ffff;
        *slot = if (state / 65536) % 100 < 10 { 1 } else { 0 };
    }

    let mut rows = vec![0i64; CELLS];
    let mut cols = vec![0i64; CELLS];
    let mut checksum: i64 = 0;

    for _ in 0..PASSES {
        for ci in 0..NCASES {
            let base = ci * CELLS;

            let mut k = 0usize;
            for r in 0..DIM {
                for c in 0..DIM {
                    if corpus[base + r * DIM + c] == 1 {
                        rows[k] = r as i64;
                        k += 1;
                    }
                }
            }

            let mut k2 = 0usize;
            for c in 0..DIM {
                for r in 0..DIM {
                    if corpus[base + r * DIM + c] == 1 {
                        cols[k2] = c as i64;
                        k2 += 1;
                    }
                }
            }

            let mut total: i64 = 0;
            if k > 0 {
                let mr = rows[k / 2];
                let mc = cols[k / 2];
                for i in 0..k {
                    total += (rows[i] - mr).abs();
                    total += (cols[i] - mc).abs();
                }
            }
            checksum = (checksum + total) % MOD;
        }
    }

    println!("checksum {checksum}");
}
